File: LEETCODE/Easy/70.ClimbingStairs.cpp
#import <vector>
#import <iostream>
#import <cassert>

class Solution {
public:
    std::vector<int> myVector;
    int DFS(int n){
            if(n < 4) return n; 
            
            // If we already calculated this step, return it immediately
            if (myVector[n] != -1) return myVector[n];   
            
            myVector[n]= (DFS(n-1)+ DFS(n-2));
            
            return myVector[n]; 
    }
    int climbStairs(int n) {
        // Initialize the vector with -1 for sizes up to n
        myVector.assign(n + 1, -1);
        return 0 + DFS(n);
    }
};


assert(Solution().climbStairs(2) == 2);
assert(Solution().climbStairs(3) == 3);
assert(Solution().climbStairs(4) == 5);
assert(Solution().climbStairs(5) == 8);

assert(Solution().climbStairs(30) == 1346269);
assert(Solution().climbStairs(45) == 1836311903);

File: LEETCODE/Easy/70.ClimbingStairs.js
const assert  = require("assert").strict;

/**
 * @param {number} n
 * @return {number}
 */
var climbStairs = function(n) {
    const DP = new Array(n+1).fill(-1);
    
    const DFS = (n) => {
        if(n < 4 ) return n;
        
        if(DP[n] != -1) return DP[n];
        
        DP[n] = DFS(n-1)+ DFS(n-2);
        
        return DP[n];
    }
    
    return DFS(n);
};


assert.strictEqual(climbStairs(2), 2);
assert.strictEqual(climbStairs(3), 3);
assert.strictEqual(climbStairs(4), 5);
//logTestcase;
assert.strictEqual(climbStairs(30), 1346269);
assert.strictEqual(climbStairs(45), 1836311903);
